### PR TITLE
JEXL expressions in custom notifications

### PR DESCRIPTION
**Pull request by fgalan**
Issue #4004

When this PR is created (at commit 9dfdccc5085b38c98bee39de4687035acb7adcb1) this PR has been locally tested with the following test cases (which are the ones that are more focused in macro replacement):

* 2560_custom_notification_json
* 2015_notification_templates
* 4085_custom_notifications_ngsi_payload
* 4159_custom_notif_extra_macros

What's pending

- [x] CI Docker
- [x] Jexl* -> Expr* (to make it more "neutral")
- [x] Python library version in `GET /version` ?
- [x] Subscription field to choose legacy/jexl evaluation (CRUD in csub) + legacy std::map based expressions mode (see [here](https://github.com/telefonicaid/fiware-orion/pull/4511#issuecomment-1980343592))
- [x] Add tranformation functions (+ ftest)
- [x] Test cases for JEXL calculations (based in different types of attributes, i.e. strings, numbers, etc.)
- [x] `TextUnrestricted` should not be needed when JEXL expression is used in `ngsi` element - 2cc242432
- [x] ~Test thread safeness model and memory profile~ (it's ok with cjexl)
- [x] ~Sem management~ (not needed with cjexl)
- [x] Use ExprResult::release() in the proper places - 49aa74c1fe2e6d770a27b375835683f03134d0ab
- [x] Add times to time counters - ecf7fbd21
    - [x] JEXL context build
    - [x] JEXL evaluation 
    - [x] ~Time waiting at sem~ (no sems at the end)
- [x] Check we have this covered in the tests:
   - [x] Unknown attribute in context - jexl_not_defined_attrs.test
   - [x] Unknown tranformation (e.g. `c|patata`) - jexl_transformation_unknown.test
   - [x] Syntax errors (eg. `c[0|fun`, `c|fun1('foo)|fun2`, `c|fun1('foo'|fun2)` - jexl_syntax_errors.test
   - [x] Use two expressions in the same notification, reusing context (to ensure the modifications to JsonHelper `str()` to allow it re-calleable are working properly) - jexl_transformation_in_id_and_type.test (`id` and `type`) and jexl_several_expressions.test (`sum` and `mul`)
- [x] Deployment Docker ~(maybe use pipx?)~
   - master (although tested using the branch for this PR) - https://github.com/telefonicaid/fiware-orion/pull/4511/commits/e789afcd7dbaae8fd4c281d3b00847f00754ca30
   - tags - https://github.com/telefonicaid/fiware-orion/pull/4511/commits/7281e1f9abd3991a01d71388781153c4a9ff51fd
- [x] Deal with cjexl lib absence
  - [x] Regression of basic replacement, see [this PR](https://github.com/telefonicaid/fiware-orion/pull/4547), solving this doubt https://github.com/telefonicaid/fiware-orion/pull/4511/commits/62d1a22ede43c3eb015ebafcfd39bc7b8fffc6eb
  - [x] Compiling - 4bf61333d
  - [x] Testing framework - d5a14947f18893c5e25e25944e8b4661057d2ea6
- [x] Analyze ftest differences (see [here](https://github.com/telefonicaid/fiware-orion/pull/4511#issuecomment-1946374835) and [here](https://github.com/telefonicaid/fiware-orion/pull/4511#issuecomment-1980343592))
  - 2015_notification_templates/notification_templates_many_notifications.test ~( + _legacy.test)~
  - 3693_covered_notifications_in_subscriptions/covered_custom_notification.test ~(+ _legacy.test)~
- [x] libcjexl 0.1.0
- [x] Check FIXME PR marks
- [x] Documentation
    - [x] orion-api.md (to explain how it works now)
    - [x] Admin (new build from sources, new CLI for the expressions mode, etc.)
    - [x] Devel (as we have created a new library directory)
    - [x] Add a warning about existing attributes in expressions with names using `-` or `:` (with basic replacement they work, but with JEXL no, see test/functionalTest/cases/4004_jexl_expressions_in_subs/jexl_expr_attrs_weird_syntax.test) - 9e04d8c5a
- [x] Functional pass with basic replacement, in separate PR https://github.com/telefonicaid/fiware-orion/pull/4552
- [x] Valgrind pass PR https://github.com/telefonicaid/fiware-orion/pull/4553
- [x] thMapper transformation (+tests)
- [x] CNR

### CODE REVIEW

**Comment by fgalan** [comment]
Trying to install pyjexl in my local (Debian) environment:

```
fermin@bodoque:~/src/fiware-orion$ sudo pip install pyjexl==0.3.0
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try apt install
    python3-xyz, where xyz is the package you are trying to
    install.
    
    If you wish to install a non-Debian-packaged Python package,
    create a virtual environment using python3 -m venv path/to/venv.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
    sure you have python3-full installed.
    
    If you wish to install a non-Debian packaged Python application,
    it may be easiest to use pipx install xyz, which will manage a
    virtual environment for you. Make sure you have pipx installed.
    
    See /usr/share/doc/python3.11/README.venv for more information.

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.

fermin@bodoque:~/src/fiware-orion$ sudo apt-get install python3-pyjexl
Leyendo lista de paquetes... Hecho
Creando árbol de dependencias... Hecho
Leyendo la información de estado... Hecho
E: No se ha podido localizar el paquete python3-pyjexl

fermin@bodoque:~/src/fiware-orion$ sudo apt-cache search pyjexl
fermin@bodoque:~/src/fiware-orion$ 
```

Virtual env is ok for Docker CI, but in regular Docker probably we should install in the system. So, let's try with `pipx`.

```
fermin@bodoque:~/src/fiware-orion$ sudo apt-get install pipx
...
fermin@bodoque:~/src/fiware-orion$ sudo pipx install pyjexl==0.3.0
Note: Dependent package 'future' contains 2 apps
  - futurize
  - pasteurize

No apps associated with package pyjexl. Try again with '--include-deps' to include apps of dependent packages, which are listed above. If you are attempting to install a library, pipx should not be used.
Consider using pip or a similar tool instead."
```

No luck so far...

**Comment by fgalan** [comment]
At commit [ae3d83b](https://github.com/telefonicaid/fiware-orion/pull/4511/commits/ae3d83b7df3f71d0572b802e4e63ebcbec4054e5) the fails happening in 2560_custom_notification_json (8) and  4085_custom_notifications_ngsi_payload (16) are due to slight differences in the JSON rendering like this ones:

imagen

imagen

This is due to the usage of `json.dumps()` in JEXL evaluation for JSON Arrays and objects and will be solved if we use the CompoundNode based rendering. To do that, the PyDic/PyList used at JEXL evaluation stage should be tranformed in a CompounNode tree. For the same price, we will avoid the call to `json.dumps()` in the Python interpreter (which seems to be a bit overkill).

With regards to the changes in this ftest

imagen

they are due to the change in behaviour when a given attribute doesn't exist in the `payload` rendering. It seems before this we use `null` and now we are using the empty string. We need to analyse this in detail (not sure if it is even an improvement...).

**Comment by fgalan** [comment]
At commit [69ff5de](https://github.com/telefonicaid/fiware-orion/pull/4511/commits/69ff5de8a7e05ad3e94295b1357d49e91ff8dcf5) we have hacked the code to regression all fuctional tests using legacy expressions:

https://github.com/telefonicaid/fiware-orion/blob/69ff5de8a7e05ad3e94295b1357d49e91ff8dcf5/src/lib/ngsiNotify/Notifier.cpp#L321-L322

An we can see the following tests failing:

  o  2015_notification_templates/notification_templates_many_notifications.test
  o  3693_covered_notifications_in_subscriptions/covered_custom_notification.test
  o  4004_jexl_expressions_in_subs/jexl_basic_attrs_in_update.test
  o  4004_jexl_expressions_in_subs/jexl_basic_attrs_not_in_update.test
  o  4004_jexl_expressions_in_subs/jexl_expr_attrs_weird_syntax.test
  o  4004_jexl_expressions_in_subs/jexl_json_navigation.test

It makes sense:

* Although 2015_notification_templates/notification_templates_many_notifications.test is failing, there is a 2015_notification_templates/notification_templates_many_notifications_legacy.test with is passing. The same for 3693_covered_notifications_in_subscriptions/covered_custom_notification.test (3693_covered_notifications_in_subscriptions/covered_custom_notification_legacy.test). We need to analyze that (see [comment](https://github.com/telefonicaid/fiware-orion/pull/4511#issuecomment-1946374835) if this is a desired behaviour in JEXL or not (noted in the checklist above)
* It makes sense that JEXL tests in 4004 are not working, due to the hack.

In sum, we can conclude that the functional test regression with legacy expressions is **OK** (so the hack can be now reverted),

**Comment by fgalan** [comment]
> ...so the hack can be now reverted

Commit 742936f9b54c2ee7a014041928e2fe659d5ef160 reverts the hack and all test are passing again.

**Comment by mapedraza** [comment]
Some tests cases to be included

## Not defined attributes:
Context: {"a":1}
Expresion: b+1
Result: Null

## Null values
Context: {"a":Null}
Expresion: a+1
Result: Null


CC: @mrutid

**Comment by fgalan** [comment]
Anther one:

Context:

```
{ a:[{b:1},{b:34},{b:4}] }
```

Expresssion:

```
a[.b<5]
```

Result:

```
[
  {
    "b": 1
  },
  {
    "b": 4
  }
]
```

**Comment by fgalan** [comment]
> Some tests cases to be included
> ## Not defined attributes:
> 
> Context: {"b":1} Expresion: a+1 Result: Null
> ## Null values
> 
> Context: {"a":Null} Expresion: a+1 Result: Null
> 
> CC: @mrutid
> 
> Edit: For the previous cases, a test with same context (value at null and undefined), but for the following expressions should be also raccomended:
> 
> !a (boolean) a+3 a*3 a|substr(0,2)
> 
> Source of inspiration: https://github.com/telefonicaid/iotagent-node-lib/blob/4c44db307f2e2723c4c8a47fdde462ced28ba754/test/functional/testCases.js#L448-L1477

Added in commit [a71c173](https://github.com/telefonicaid/fiware-orion/pull/4511/commits/a71c173a3a9092b42d5df85e9bc61c9ce0311a28)

**Comment by fgalan** [comment]
At the end we are not going to implement the exprLang selector. Removed in commit https://github.com/telefonicaid/fiware-orion/pull/4511/commits/9434b9c95a6a9a644f25fa7b2de41fc5bda01bbe (which is a kind of revert of previous commits https://github.com/telefonicaid/fiware-orion/pull/4511/commits/92e92896358aaa61e150581758862f5562e6713f and https://github.com/telefonicaid/fiware-orion/commit/238667780e62055d18643c4f9b18f3de9c5c11f4, with some extra adjustements)

File: test/functionalTest/cases/4004_jexl_expressions_in_subs/jexl_null_values.test [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+orionCurl --url /v2/entities --payload "$payload"
+echo
+echo
+
+
+echo "03. Update entity E1 with A=null again (with forcedUpdate)"
+echo "=========================================================="
+payload='{
+  "A": {
+    "value": "null",
```
[the remark this answers is not in the document]
Comment (fgalan) [reply test/functionalTest/cases/4004_jexl_expressions_in_subs/jexl_null_values.test:89]: This step is supossed to be the same as step 02, so it isn't intentional...

[the remark this answers is not in the document]
Comment (fgalan) [reply test/functionalTest/cases/4004_jexl_expressions_in_subs/jexl_null_values.test:89]:
```suggestion
    "value": null,
```

File: .github/workflows/publishimage-master.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+name: Publish Docker image (master)
+
+# The workflow will push images for master on every merge
```
Comment (fgalan) [inline .github/workflows/publishimage-master.yml:3]:
```suggestion
# The workflow will push images for master on every merge
# Ideally, this should be done at dockerhub, but it doesn't support secrets (see https://stackoverflow.com/questions/78446824/use-environment-variables-with-sensible-information-in-docker-hub-autobuild)
```

File: .github/workflows/publishimage-tag.yml [hunk 3]
```diff
@@ -0,0 +1,35 @@
+name: Publish Docker image (tag)
+
+# The workflow will push images on every tag in the format x.y.z
```
Comment (fgalan) [inline .github/workflows/publishimage-tag.yml:3]:
```suggestion
# The workflow will push images on every tag in the format x.y.z
# Ideally, this should be done at dockerhub, but it doesn't support secrets (see https://stackoverflow.com/questions/78446824/use-environment-variables-with-sensible-information-in-docker-hub-autobuild)
```

File: CHANGES_NEXT_RELEASE [hunk 4]
```diff
@@ -1,2 +1,7 @@
+- Add: JEXL expression support in custom notification macro replacement (using cjexl 0.2.0) (#4004)
+- Add: expression context build and evaluation counters in timing section in GET /statistics (#4004)
+- Fix: use null for non existing attributes in custom covered notifications macro substitution (instead of empty string) to make behaviour more consistent (#4004)
+- Fix: use null for non existing attributes in macro substitution applied to "paylaod" field (instead of empty string) to make behaviour more consistent (#4004)
 - Fix: simplified GET /version operation, without including "libversions" field (add ?options=libVersions to get it)
 - Fix: lighter operation to get databases list from MongoDB (#4517)
+
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE:7]:
```suggestion
```

File: doc/manuals/orion-api.md [hunk 5]
```diff
@@ -398,11 +427,12 @@ the following will be used (note that "%25" is the encoding for "%").
 GET /v2/entities/E%253C01%253E
 ```
 
-There are some exception cases in which the above restrictions do not apply. In particular, in the following fields:
+There are some exception cases in which the above restrictions do not apply. In particular, in the following cases:
 
 * URL parameter `q` allows the special characters needed by the [Simple Query Language](#simple-query-language)
 * URL parameter `mq` allows the special characters needed by the [Simple Query Language](#simple-query-language)
 * URL parameter `georel` and `coords` allow `;`
+* Within `ngsi` (i.e. `id`, `type` and attribute values) in [NGSI Payload patching](#ngsi-payload-patching) (to support characters used in the JEXL expression syntax)
```
Comment (fgalan) [inline doc/manuals/orion-api.md:435]:
```suggestion
* Within `ngsi` (i.e. `id`, `type` and attribute values) in [NGSI Payload patching](#ngsi-payload-patching) (to support characters used in the [JEXL expression syntax](#jexl-support))
```

File: src/lib/expressions/ExprContext.cpp [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+
+
+
+/* ****************************************************************************
+*
+* ExprContextList::add -
+*/
+void ExprContextList::add(ExprContextList exprContextList)
+{
+
```
Comment (fgalan) [inline src/lib/expressions/ExprContext.cpp:282]:
```suggestion
```

File: src/lib/expressions/ExprContext.cpp [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+
+
+/* ****************************************************************************
+*
+* ExprContextList::add -
+*/
+void ExprContextList::add(const std::string &_value)
+{
+
```
Comment (fgalan) [inline src/lib/expressions/ExprContext.cpp:220]:
```suggestion
```

File: src/lib/expressions/ExprManager.cpp [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+/* ****************************************************************************
+*
+* ExprManager::init -
+*/
+void ExprManager::init(void)
+{
+  jexlEngine = cjexl_new_engine();
+}
+
```
Comment (fgalan) [inline src/lib/expressions/ExprManager.cpp:72]:
```suggestion
```

File: src/lib/rest/rest.cpp [hunk 9]
```diff
@@ -618,6 +618,10 @@ static void requestCompleted
     clock_addtime(&accTimeStat.mongoWriteWaitTime,    &threadLastTimeStat.mongoWriteWaitTime);
     clock_addtime(&accTimeStat.mongoReadWaitTime,     &threadLastTimeStat.mongoReadWaitTime);
     clock_addtime(&accTimeStat.mongoCommandWaitTime,  &threadLastTimeStat.mongoCommandWaitTime);
+    clock_addtime(&accTimeStat.exprBasicCtxBldTime,  &threadLastTimeStat.exprBasicCtxBldTime);
+    clock_addtime(&accTimeStat.exprBasicEvalTime,    &threadLastTimeStat.exprBasicEvalTime);
```
Comment (fgalan) [inline src/lib/rest/rest.cpp:622]:
```suggestion
    clock_addtime(&accTimeStat.exprBasicCtxBldTime,   &threadLastTimeStat.exprBasicCtxBldTime);
    clock_addtime(&accTimeStat.exprBasicEvalTime,     &threadLastTimeStat.exprBasicEvalTime);
```

File: src/lib/expressions/ExprManager.cpp [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+static void cjexl_free_engine(void* ptr)
+{
+}
+
+static const char* cjexl_eval(void* ptr, const char* script_ptr, const char* context_ptr)
+{
+  return "";
+}
+#else
+// Interface to use libcjexl.a
```
Comment (fgalan) [inline src/lib/expressions/ExprManager.cpp:49]:
```suggestion
// Interface to use libcjexl
```

File: src/lib/serviceRoutines/versionTreat.cpp [hunk 11]
```diff
@@ -47,6 +47,13 @@
 #include <bson/bson.h>
 #include <mosquitto.h>
 
+#ifndef EXPR_BASIC
+// Interface to use libcjexl.a
```
Comment (fgalan) [inline src/lib/serviceRoutines/versionTreat.cpp:51]:
```suggestion
// Interface to use libcjexl
```

File: doc/manuals/orion-api.md [hunk 12]
```diff
@@ -2295,6 +2326,676 @@ Some considerations to take into account when using custom notifications:
...
+```
+"medium"
+```
+
+### Failsafe cases
+
+As failsafe behaviour, evaluation returns `null` in the following cases:
+
+* Some of the transformation used in the expression is unknown
+* Operations with identifiers that are not defined in the context are used. For instance, `(A==null)?-1:A` will result in `null` (and not `-1`) if `A` is not in the context, due to `==` is an operation that cannot be done on undefined identifiers. However, `A||-1` will work (i.e. `-1` will result if `A` is not in the context), as `||` is not considered an operation on `A`.
```
Comment (danielvillalbamota) [inline doc/manuals/orion-api.md:2991]:
```suggestion
* Operations with identifiers that are not defined in the context are used. For instance, `(A==null)?1:A` will result in `null` (and not `1`) if `A` is not in the context, due to `==` is an operation that cannot be done on undefined identifiers. However, `A||1` will work (i.e. `1` will result if `A` is not in the context), as `||` is not considered an operation on `A`.
```

As `A||-1` include minus operator and it is a "Known limitation", I suggest using this example

Comment (fgalan) [reply doc/manuals/orion-api.md:2991]:
```suggestion
* Operations with identifiers that are not defined in the context are used. For instance, `(A==null)?0:A` will result in `null` (and not `0`) if `A` is not in the context, due to `==` is an operation that cannot be done on undefined identifiers. However, `A||0` will work (i.e. `0` will result if `A` is not in the context), as `||` is not considered an operation on `A`.
```

Comment (fgalan) [reply doc/manuals/orion-api.md]: Fixed in [fbde34e](https://github.com/telefonicaid/fiware-orion/pull/4511/commits/fbde34e73bc50a01e63a11b9473a295a7723262a)

File: test/functionalTest/cases/4004_jexl_expressions_in_subs/jexl_null_values_all_cases.test [hunk 13]
```diff
@@ -0,0 +1,282 @@
...
+# VALGRIND_READY - to mark the test ready for valgrindTestSuite.sh
+# JEXL_EXPR_FLAVOUR - to mark the test has to execute only when contextBroker includes jexl-expr flavour
+
+--NAME--
+Expression in custom notification (null values in all cases with JEXL expression fail)
+
+--SHELL-INIT--
+dbInit CB
+brokerStart CB
+accumulatorStart
```
Comment (danielvillalbamota) [inline test/functionalTest/cases/4004_jexl_expressions_in_subs/jexl_null_values_all_cases.test:30]:
```suggestion
accumulatorStart --pretty-print
```

Comment (fgalan) [reply test/functionalTest/cases/4004_jexl_expressions_in_subs/jexl_null_values_all_cases.test:30]: Exactly :)

Comment (danielvillalbamota) [reply test/functionalTest/cases/4004_jexl_expressions_in_subs/jexl_null_values_all_cases.test:30]: ok, NTC

File: test/functionalTest/cases/4004_jexl_expressions_in_subs/jexl_transformation_full.test [hunk 14]
```diff
@@ -0,0 +1,450 @@
...
+          "type": "Text",
+          "value": "${K|typeOf}"
+        },
+        "L": {
+          "type": "TextUnrestricted",
+          "value": "${(L|split('\'' '\''))}"
+        },
+        "M": {
+          "type": "Number",
+          "value": "${M|round(2)}"
```
Comment (danielvillalbamota) [inline test/functionalTest/cases/4004_jexl_expressions_in_subs/jexl_transformation_full.test:131]: `round` does not receive arguments, is it this expression intentional?

Comment (fgalan) [reply test/functionalTest/cases/4004_jexl_expressions_in_subs/jexl_transformation_full.test:131]: No, it's a typo. Let's fix it an check the test is still passing

```suggestion
          "value": "${M|round}"
```

(Waiting to finish the current test pass to commit this suggestion...)

Comment (fgalan) [reply test/functionalTest/cases/4004_jexl_expressions_in_subs/jexl_transformation_full.test]: Fixed in [cfa543a](https://github.com/telefonicaid/fiware-orion/pull/4511/commits/cfa543a5408363e6be32895bd27fe4b5d21a8d68)

**Review by danielvillalbamota — APPROVED** [review]
LGTM

File: doc/manuals/orion-api.md [hunk 15]
```diff
@@ -2295,6 +2326,676 @@ Some considerations to take into account when using custom notifications:
...
+
+```
+"medium"
+```
+
+### Failsafe cases
+
+As failsafe behaviour, evaluation returns `null` in the following cases:
+
+* Some of the transformation used in the expression is unknown
```
Comment (mapedraza) [inline doc/manuals/orion-api.md:2990]: Probably, is a good idea to add an example:

```suggestion
* Some of the transformation used in the expression is unknown (e.g. `A|undefinedExpression`)
```

Comment (fgalan) [reply doc/manuals/orion-api.md]: Good idea!

Fixed in [06cbe4e](https://github.com/telefonicaid/fiware-orion/pull/4511/commits/06cbe4e3506c56a67c81207a86cd29555dd58fdf)

**Review by mapedraza — APPROVED** [review]
LGTM

**Comment by fgalan** [comment]
@fisuda this PR does some modifications in the .md of the English documentation. It would be great if you could do a PR with the corresponding Japanese modifications. Thanks in advance!

**Comment by fisuda** [comment]
I sent the PR #4559. Thanks.

**Pull request merged**